### ISSUE TITLE
[java] CloseResource false positive

### ISSUE THREAD

**Issue by ben-manes** [pmd/pmd#1911]
**Affects PMD Version:** 6.16
**Rule:** [CloseResource](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#closeresource)
**Description:**
Sometimes code flows of many closeables is harder to express using nested try-with-resource blocks. Instead a Guava `Closer` can be used to aggregate the resources and close them, as in a single try-with-resource block. This is particularly useful when libraries don't implement `AutoCloseable` or to pass the aggregator to methods for resources that have to be closed at the end of the entire block. 

In a case like below, I am receiving a `CloseResource` warning for `SequenceWriter`. Would it be possible to suppose `Closer` specifically if it cannot be deduced otherwise?

```
jsonSchema/csv/CsvToJson.java | 119 | Ensure that resources like this SequenceWriter object are closed after use
```

**Code Sample demonstrating the issue:**
```java
try (Closer closer = Closer.create()) {
  CsvParser parser = new CsvParser(settings);
  closer.register(parser::stopParsing);
  parser.beginParsing(reader);
  ...
  SequenceWriter output = closer.register(mapper.writerFor(Entity.class)
      .withDefaultPrettyPrinter()
      .writeValuesAsArray(writer));
  while ((record = parser.parseNextRecord()) != null) {
    output.write(recordToEntity(record));
  }
} catch (IOException e) {
  throw new UncheckedIOException(e);
}
```

**Comment by ben-manes** [comment]
Another case I ran into,
```
ETLService.java	143	Ensure that resources like this Connection object are closed after use
ETLService.java	144	Ensure that resources like this Statement object are closed after use
ETLService.java	145	Ensure that resources like this ResultSet object are closed after use
```

```java
Connection conn = null;
Statement stmt = null;
ResultSet rs = null;
try {
  ...
} finally {
  JdbcUtils.closeResultSet(rs);
  JdbcUtils.closeStatement(stmt);
  JdbcUtils.closeConnection(conn);
}
```

**Comment by jsotuyod** [comment]
> Another case I ran into,
> 
> ```
> ETLService.java	143	Ensure that resources like this Connection object are closed after use
> ETLService.java	144	Ensure that resources like this Statement object are closed after use
> ETLService.java	145	Ensure that resources like this ResultSet object are closed after use
> ```
> 
> ```java
> Connection conn = null;
> Statement stmt = null;
> ResultSet rs = null;
> try {
>   ...
> } finally {
>   JdbcUtils.closeResultSet(rs);
>   JdbcUtils.closeStatement(stmt);
>   JdbcUtils.closeConnection(conn);
> }
> ```

The `closeTargets` property is designed specifically to address these scenarios, simply set it to `JdbcUtils.closeResultSet,JdbcUtils.closeStatement,JdbcUtils.closeConnection`

**Comment by ben-manes** [comment]
Oh thank you, I completely missed `closeTargets`.

**Comment by jsotuyod** [comment]
@ben-manes no problem!

As for the original report… I don't think this is something PMD can actually handle on it's own, as it implies knowing the semantics / internals of Guava's `Closer`.

Unless someone comes up with something incredibly smart, this seems like a case you should simply whitelist with a `@SuppressWarnings("PMD.CloseResource")` or `// NOPMD.`

**Comment by ben-manes** [comment]
yes, I'm not surprised by that. I was hoping there would be a way to instruct the rule to assume a close if the resource was passed into a whitelisted aggregator object (`Collecting Parameter` in refactorings speak). That way if someone later added a new resource and forgot to include it, the warning would still be caught. But that might also be too magical of a requested feature.

**Comment by marcopelegrini** [comment]
Fun one:
`Ensure that resources like this boolean object are closed after use`

Because of:

```
final Stream<V1Container> containers = spec.getContainers() != null ? spec.getContainers().stream() : Stream.empty();
```

How the hell do I close a boolean? :)

**Comment by ben-manes** [comment]
Closing as I don’t think my ask is reasonable or viable
